crates/kornia-3d/src/voxel_grid.rs

```diff
@@ -0,0 +1,382 @@
+    ///
+    /// # Arguments
+    /// * `leaf_size` - The size of the voxel in x, y, z dimensions.
+    ///
+    /// # Panics
+    /// Panics if any dimension of `leaf_size` is non-positive.
+    pub fn new(leaf_size: DVec3) -> Self {
+        if leaf_size.x <= 0.0 || leaf_size.y <= 0.0 || leaf_size.z <= 0.0 {
+            panic!("Leaf size must be positive in all dimensions");
+        }
+
+        VoxelGrid {
+            leaf_size,
+            min_points_per_voxel: 1,
+            downsample_all_data: true,
+        }
+    }
+
+    /// Downsamples the input point cloud by grouping points into voxels and computing centroids.
+    ///
+    /// # Arguments
+    /// * `point_cloud` - The input point cloud to downsample.
+    ///
+    /// # Returns
+    /// A new `PointCloud` containing the downsampled points, and optionally colors and normals.
+    pub fn downsample(&mut self, point_cloud: &PointCloud) -> PointCloud {
+        let mut grid: HashMap<(i32, i32, i32), VoxelData> = HashMap::new();
+
+        // Group points into voxels
+        for (i, point) in point_cloud.points().iter().enumerate() {
+            let point_vec = DVec3::from_array(point);
+            let ix = (point_vec.x / self.leaf_size.x).floor() as i32;
+            let iy = (point_vec.y / self.leaf_size.y).floor() as i32;
+            let iz = (point_vec.z / self.leaf_size.z).floor() as i32;
+            let key = (ix, iy, iz);
+            let entry = grid.entry(key).or_insert((
+                DVec3::from_array(&[0.0, 0.0, 0.0]),
+                Vec::new(),
+                Vec::new(),
+                0,
+            ));
+            entry.0.x += point_vec.x;
+            entry.0.y += point_vec.y;
+            entry.0.z += point_vec.z;
+            entry.3 += 1;
+
+            if self.downsample_all_data {
+                if let Some(colors) = point_cloud.colors() {
+                    if let Some(color) = colors.get(i) {
+                        entry.1.push(*color);
+                    }
+                }
+                if let Some(normals) = point_cloud.normals() {
+                    if let Some(normal) = normals.get(i) {
+                        entry.2.push(*normal);
+                    }
+                }
+            }
+        }
+
+        // Compute centroids for each voxel
+        let mut points = Vec::new();
+        let mut colors = if self.downsample_all_data && point_cloud.colors().is_some() {
+            Some(Vec::new())
+        } else {
+            None
+        };
+        let mut normals = if self.downsample_all_data && point_cloud.normals().is_some() {
+            Some(Vec::new())
+        } else {
+            None
+        };
+
+        for (_key, (sum, color_vec, normal_vec, count)) in grid {
+            if count >= self.min_points_per_voxel {
+                let inv_count = 1.0 / count as f64;
+                let centroid =
+                    DVec3::from_array(&[sum.x * inv_count, sum.y * inv_count, sum.z * inv_count]);
```

Comment:
```suggestion
                    sum * inv_count;
```
that should work i think